circuit/src/layered/circuit.rs

```diff
@@ -142,9 +143,25 @@ impl<C: GKRFieldConfig> Clone for Circuit<C> {
 unsafe impl<C> Send for Circuit<C> where C: GKRFieldConfig {}
 
 impl<C: GKRFieldConfig> Circuit<C> {
-    pub fn load_circuit(filename: &str) -> Self {
+    pub fn load_circuit<Cfg: GKRConfig<FieldConfig = C>>(filename: &str) -> Self {
         let rc = RecursiveCircuit::<C>::load(filename).unwrap();
-        rc.flatten()
+        let mut circuit = rc.flatten();
+
+        circuit.identify_rnd_coefs();
+        circuit.identify_structure_info();
+
+        // If there will be two claims for the input
+        // Introduce an extra relay layer before the input layer
+        if !circuit.layers[0].structure_info.skip_sumcheck_phase_two {
+            match Cfg::PCS_TYPE {
+                // Raw PCS costs nothing in opening, so no need to add relay layer
+                // But we can probably add it in the future for verifier's convenience
+                config::PolynomialCommitmentType::Raw => (),
+                _ => circuit.add_input_relay_layer(),
+            }
+        }
```

Comment:
I have a random thought, that can we just clean up the information of skipping phase 2, and on default only do 1 PCS opening?  An additional layer of sumcheck would be generally cheaper than a PCS opening, but I understand that removing the condition of skipping phase 2 means a few changes on recursive side and prover side, so we can put it as is by now.